boot/bootutil/include/bootutil/fault_injection_hardening.h

```diff
@@ -0,0 +1,339 @@
+{
+    return x;
+}
+
+__attribute__((always_inline)) inline
+int fih_eq(fih_int x, fih_int y)
+{
+    return x == y;
+}
+__attribute__((always_inline)) inline
+int fih_not_eq(fih_int x, fih_int y)
+{
+    return x != y;
+}
+#endif /* FIH_ENABLE_DOUBLE_VARS */
+
+/* C has a common return pattern where 0 is a correct value and all others are
+ * errors. This function converts 0 to FIH_SUCCESS and any other number to a
+ * value that is not FIH_SUCCESS
+ */
+__attribute__((always_inline)) inline
+fih_int fih_int_encode_zero_equality(int x)
+{
+    if (x) {
+        return FIH_FAILURE;
+    } else {
+        return FIH_SUCCESS;
+    }
+}
+
+#ifdef FIH_ENABLE_CFI
+extern fih_int _fih_cfi_ctr;
+#endif /* FIH_ENABLE_CFI */
+
+fih_int fih_cfi_get_and_increment(void);
+void fih_cfi_validate(fih_int saved);
+void fih_cfi_decrement(void);
+
+/* Label for interacting with FI testing tool. Can be parsed from the elf file
+ * after compilation. Does not require debug symbols.
+ */
+#define FIH_LABEL(str) __asm volatile ("FIH_LABEL_" str "_%=:" ::);
+
+/* Main FIH calling macro. return variable is second argument. Does some setup
+ * before and validation afterwards. Inserts labels for use with testing script.
+ *
+ * First perform the precall step - this gets the current value of the CFI
+ * counter and saves it to a local variable, and then increments the counter.
+ *
+ * Then set the return variable to FIH_FAILURE as a base case.
+ *
+ * Then perform the function call. As part of the funtion FIH_RET must be called
+ * which will decrement the counter.
+ *
+ * The postcall step gets the value of the counter and compares it to the
+ * previously saved value. If this is equal then the function call and all child
+ * function calls were performed.
+ */
+#define FIH_CALL(f, ret, ...) \
+    do { \
+        FIH_LABEL("FIH_CALL_START"); \
+        FIH_CFI_PRECALL_BLOCK; \
+        ret = FIH_FAILURE; \
+        if (fih_delay()) { \
+            ret = f(__VA_ARGS__); \
+        } \
+        FIH_CFI_POSTCALL_BLOCK; \
+        FIH_LABEL("FIH_CALL_END"); \
+    } while (0)
+
+/* FIH return changes the state of the internal state machine. If you do a
+ * FIH_CALL then you need to do a FIH_RET else the state machine will detect
+ * tampering and panic.
+ */
+#define FIH_RET(ret) \
+    do { \
+        FIH_CFI_PRERET; \
+        return ret; \
+    } while (0)
+
+/* Macro wrappers for functions - Even when the functions have zero body this
+ * saves a few bytes on noop functions as it doesn't generate the call/ret
+ *
+ * CFI precall function saves the CFI counter and then increments it - the
+ * postcall then checks if the counter is equal to the saved value. In order for
+ * this to be the case a FIH_RET must have been performed inside the called
+ * function in order to decrement the counter, so the function must have been
+ * called.
+ */
+#ifdef FIH_ENABLE_CFI
+#define FIH_CFI_PRECALL_BLOCK \
+    fih_int _fih_cfi_saved_value = fih_cfi_get_and_increment()
```

Comment:
Just some blue sky thinking, not sure whether it can be done or not. If the increment can be done based on the function being invoked i.e some way to hash the name of the function into a unique number, then we can confirm if an increment is called for a function call, then the decrement in the same function need to be invoked for it to be validated. 

This would ensure that if the call is subverted to a different location and the RET from that different function will fail the check on return back to the caller.

Reply:
It would be an interesting idea, but maybe too difficult in practice. The C preprocessor does have access to the function name via `__func__`, but I'm not sure if you could evaluate a hash function or similar inside the preprocessor. I'd avoid doing the hash inside actual code, as that'd cause a fairly large size increase due to the inlining.

In general, most functions don't have an FIH_RET and it's pretty hard to control the value of register corruption (and hence jump location). More targeted attacks like ion beams could probably target the lower bits deliberately, which would allow jumps to be slightly more precise.

Reply:
This would be ideal. A similar alternative might be easier: take the function address and derive a number from it. 

It's worth noting that fault injection techniques are unlikely to produce precise values for a return address. If precise return addresses are possible then it is time to consider more comprehensive ROP/JOP protection, which I believe is beyond the scope of the attacks/adversaries considered in this PR.

Reply:
I see Raef beat me to it :)

Reply:
Might `sizeof(__func__)` would add some randomness.